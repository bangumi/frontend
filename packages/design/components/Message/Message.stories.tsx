import type { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';

import Message from '.';

const componentMeta: ComponentMeta<typeof Message> = {
  title: 'modern/Message',
  component: Message,
  parameters: {
    docs: {
      description: {
        component: '消息提示块，可以直接用于页面中，也用于Toast组件中。',
      },
    },
  },
  argTypes: {
    children: {
      description: '消息内容',
      defaultValue: '这是一条消息提示',
      control: { type: 'text' },
      table: {
        type: { summary: 'string' },
      },
    },
    type: {
      description: '消息类型：对应不同样式',
      control: { type: 'select' },
      options: ['info', 'error'],
      table: {
        type: { summary: 'string' },
        defaultValue: { summary: 'info' },
      },
      defaultValue: 'info',
    },
    blockWidth: {
      description: '消息长度：是否占据整行',
      control: { type: 'inline-radio' },
      options: [true, false],
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: false },
      },
      defaultValue: false,
    },
  },
  decorators: [
    (story) => (
      <div
        style={{
          width: '500px',
          height: '200px',
        }}
      >
        {story()}
      </div>
    ),
  ],
};
export default componentMeta;

const Template: ComponentStory<typeof Message> = (args) => {
  return (
    <>
      <Message {...args} />
    </>
  );
};

export const Usage = Template.bind({});
