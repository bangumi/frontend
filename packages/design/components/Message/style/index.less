@import '../../../theme/base';

.bgm-message {
  display: flex;

  &__content {
    border-radius: 19px;
    padding: 9px 16px;
    box-sizing: border-box;
    display: flex;
    align-items: center;
    font-weight: 600;
    font-size: 14px;
    line-height: 20px;

    > svg {
      flex: none;
      width: 20px;
      height: 20px;
      margin-right: 8px;
    }
  }

  &__content--block {
    flex: 1;
  }

  &__content--info {
    background-color: rgba(90, 87, 87, 0.6);
    color: #fff;
  }

  &__content--error {
    background-color: #f97f77;
    color: #fff;
  }
}
