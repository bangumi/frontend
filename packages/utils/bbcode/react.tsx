import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
