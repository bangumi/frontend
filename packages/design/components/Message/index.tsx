import './style';

import classnames from 'classnames';
import React from 'react';

import { Error } from '@bangumi/icons';

export type MessageType = 'info' | 'error';
export interface MessageProps {
  /** 消息类型：对应不同样式 */
  type?: MessageType;
  /** 消息长度：是否占据整行 */
  blockWidth?: boolean;
  /** 自定义类名 */
  className?: string;
}

const Message = ({
  type = 'info',
  blockWidth = false,
  className,
  children,
}: React.PropsWithChildren<MessageProps>) => {
  const classes = classnames(
    'bgm-message__content',
    `bgm-message__content--${type}`,
    { 'bgm-message__content--block': blockWidth },
    className,
  );

  return (
    <div className='bgm-message'>
      <div className={classes}>
        <Error />
        <span>{children}</span>
      </div>
    </div>
  );
};

export default Message;
