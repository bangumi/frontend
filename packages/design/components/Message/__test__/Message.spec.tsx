import { render } from '@testing-library/react';
import React from 'react';

import Message from '..';

describe('<Message />', () => {
  it('should be render a Message', () => {
    const { container } = render(<Message>example-message</Message>);
    expect(container).toMatchSnapshot();
  });

  it('should be a whole line', () => {
    const { container } = render(<Message blockWidth>example-message-block</Message>);
    expect(container).toMatchSnapshot();
  });

  it('should be a error message', () => {
    const { container } = render(<Message type='error'>example-message-error</Message>);
    expect(container).toMatchSnapshot();
  });
});
