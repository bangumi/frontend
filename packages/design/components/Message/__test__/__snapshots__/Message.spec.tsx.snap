// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Message /> > should be a error message 1`] = `
<div>
  <div
    class="bgm-message"
  >
    <div
      class="bgm-message__content bgm-message__content--error"
    >
      <div />
      <span>
        example-message-error
      </span>
    </div>
  </div>
</div>
`;

exports[`<Message /> > should be a whole line 1`] = `
<div>
  <div
    class="bgm-message"
  >
    <div
      class="bgm-message__content bgm-message__content--info bgm-message__content--block"
    >
      <div />
      <span>
        example-message-block
      </span>
    </div>
  </div>
</div>
`;

exports[`<Message /> > should be render a Message 1`] = `
<div>
  <div
    class="bgm-message"
  >
    <div
      class="bgm-message__content bgm-message__content--info"
    >
      <div />
      <span>
        example-message
      </span>
    </div>
  </div>
</div>
`;
